Add POSDAO transition and malice report queue.

afck opened the pull request:
This adds a `posdaoTransition` parameter to enable:

* Calling the `emitInitiateChange` function of the validator set contract.
The [POSDAO](https://forum.poa.network/t/posdao-white-paper/2208) network needs one of the validators to call this, so that a log event is added to a block when the validator set changes.
* A malice report queue.
Misbehaving validators can be reported to the validator set contract, by making a transaction with a call to `reportMalicious`. However, transactions may fail to get included and may even become obsolete, if some other transaction was made with their nonce. The queue keeps track of malice reports, and retries them with updated nonces until they succeed.
* Setting the gas price for malice reports to zero: Validator nodes shouldn't need to have positive balance in their signing accounts.

This is based on https://github.com/poanetwork/parity-ethereum/pull/151, https://github.com/poanetwork/parity-ethereum/pull/129, and https://github.com/poanetwork/parity-ethereum/commit/424c593932a273c09aea2243f4b7fcf66d6e1388.

It shares some code with https://github.com/paritytech/parity-ethereum/pull/10946.

afck commented:
Those test failures seem to be problems with the CI setup again.

dvdplm commented on ethcore/engines/validator-set/src/safe_contract.rs:
```diff
@@ -300,6 +383,88 @@ impl ValidatorSet for ValidatorSafeContract {
+		let mut queue = self.queued_reports.lock().unwrap();
+
+		if queue.len() > MAX_QUEUED_REPORTS {
+			warn!(target: "engine", "Removing report from report cache, even though it has not been finalized");
```
```suggestion
			warn!(target: "engine", "Removing {} reports from report cache, even though it has not been finalized", queue.len() - MAX_QUEUED_REPORTS);
```
maybe?

dvdplm commented on ethcore/engines/validator-set/src/safe_contract.rs:
```diff
@@ -283,6 +317,55 @@ impl ValidatorSafeContract {
+		let client = client.as_full_client().ok_or("No full client!")?;
+
+		self.queued_reports.lock().unwrap().retain(|&(malicious_validator_address, block, ref _data)| {
+			trace!(target: "engine", "Checking if report can be removed from cache");
```
```suggestion
			trace!(target: "engine", "Checking if report should be removed from cache");
```
?
And maybe include a way to know which report we're checking here? Include the malicious validator address and the block?

dvdplm commented on ethcore/engines/validator-set/src/safe_contract.rs:
```diff
@@ -300,6 +383,88 @@ impl ValidatorSet for ValidatorSafeContract {
+		}
+
+		// Retry all pending reports.
+		self.filter_report_queue(header.author(), header.number())?;
```
How come `filter_report_queue()` isn't implemented as a member of the `queued_reports` itself? So the line below could be something like `let queued_reports = self.queued_reports.lock().unwrap().filter_reports()` (or something to the same effect)?

afck replied:
I extracted the report queue into a new struct; please let me know whether that's what you had in mind. (I put it in a separate commit, so we can revert it if you don't agree.)

dvdplm commented on ethcore/engines/validator-set/src/safe_contract.rs:
```diff
@@ -283,6 +317,55 @@ impl ValidatorSafeContract {
+			if latest_block > 100 && latest_block - 100 > block {
+				return false; // Report is too old and cannot be used
+			}
+			// Check if the validator is already banned...
```
Just thinking out aloud here, but isn't it worthwhile to cut down on the number of contract calls? Could it make a single call like `does_this_guy_need_to_be_reported::call()? In other words: it doesn't seem like we care about the reason a report should be dropped.
Taking that one step further: would it be more performant to just send off all the reports we have and let the contract decide if the report is a duplicate or already processed?

afck replied:
Sure, I'd be happy to replace them with a single call per report, or even with a single call for the whole list. I don't know how expensive constant local contract calls are, and what the overhead of several simple vs. a single complex call is.

@varasev: What do you think about this API change?

varasev replied:
OK, I added a separate getter to `ValidatorSet` contract, so let's replace [these lines](https://github.com/poanetwork/parity-ethereum/blob/dbdd9c5f0f0cbe274b7cc6962decc997e3ebe319/ethcore/engines/validator-set/src/safe_contract.rs#L592-L629) with a single call to [that getter](https://github.com/poanetwork/posdao-contracts/blob/2700b23fb4e6ad2a80802c72175c2a18b2149cb7/contracts/ValidatorSetAuRa.sol#L562-L586).

afck replied:
Done.

dvdplm commented on ethcore/engines/validator-set/src/safe_contract.rs:
```diff
@@ -47,6 +49,11 @@ use super::simple_list::SimpleList;
 
+/// The maximum number of reports to keep queued.
+const MAX_QUEUED_REPORTS: usize = 10;
+/// The maximum number of returned malice reports to resend.
```
I don't understand this comment. Can you expand/elaborate a bit?

dvdplm commented on ethcore/engines/validator-set/src/contract.rs:
```diff
@@ -48,31 +49,64 @@ pub struct ValidatorContract {
+
+		/// Returns the gas price for report transactions.
+		///
+		/// After `posdaoTransition`, this is zero. Otherwise it is the default (`None`).
```
whitespace

dvdplm commented on ethcore/engines/validator-set/src/contract.rs:
```diff
@@ -48,31 +49,64 @@ pub struct ValidatorContract {
+		}
+	}
+
+	fn do_report_malicious(&self, address: &Address, block: BlockNumber, proof: Bytes) -> Result<(), EthcoreError> {
```
Maybe?
```suggestion
	fn enqueue_malicious_behaviour_report(&self, address: &Address, block: BlockNumber, proof: Bytes) -> Result<(), EthcoreError> {
```

afck replied:
But we not only enqueue, but also send a transaction:
```rust
self.transact(data, gas_price)?;
```
The queue is only a fallback, so that we retry if the transaction fails to get included in a block.

varasev commented on ethcore/engines/authority-round/src/lib.rs:
```diff
@@ -583,6 +588,9 @@ pub struct AuthorityRound {
 	/// Memoized gas limit overrides, by block hash.
 	gas_limit_override_cache: Mutex<LruCache<H256, Option<U256>>>,
+	/// The block number at which the consensus engine switches from AuRa to AuRa with POSDAO
+	/// modifications.
```
The link could be https://www.xdaichain.com/for-validators/posdao-whitepaper

afck commented:
Sorry for the indentation confusion! (Trying with vim-IndentConsistencyCop now.)
I tried to address all comments so far.

afck commented:
I can't reproduce those test failures locally.

dvdplm commented:
> I can't reproduce those test failures locally.

It's a flaky test, apologies. Restarted the job.

varasev commented:
Hi all - It seems all suggestions have been addressed. Are there any other proposals we could take into account here? Or this PR could be approved?

dvdplm commented on ethcore/engines/validator-set/src/safe_contract.rs:
```diff
@@ -200,14 +211,37 @@ fn prove_initial(contract_address: Address, header: &Header, caller: &Call) -> R
 		}
 	}
 
+	pub(crate) fn queue_report(&self, addr: Address, block: BlockNumber, data: Vec<u8>) {
```
```suggestion
	pub(crate) fn enqueue_report(&self, addr: Address, block: BlockNumber, data: Vec<u8>) {
```

varasev replied:
Done in https://github.com/paritytech/parity-ethereum/pull/11245/commits/f7c6325a9a0f7cc9bbe4000a9e6a2778ccb8129f

ordian commented on ethcore/engines/authority-round/src/lib.rs:
```diff
@@ -1116,6 +1126,56 @@ impl AuthorityRound {
+			None => return Ok(Vec::new()), // We are not a validator, so we shouldn't call the contracts.
+		};
+		let our_addr = signer.address();
+		let client = self.client.read().as_ref().and_then(|weak| weak.upgrade()).ok_or_else(|| {
```
should it use the newly added [`upgrade_client_or`](https://github.com/paritytech/parity-ethereum/pull/11408/files#diff-79ef4a3dce58a9d303d49c5aee781e17R1124) method?

ordian commented on ethcore/engines/validator-set/src/safe_contract.rs:
```diff
@@ -16,9 +16,10 @@
 
-use std::sync::{Weak, Arc};
+use std::collections::VecDeque;
+use std::sync::{Arc, Mutex, Weak};
```
why not `parking_lot::Mutex`? we're using `parking_lot::RwLock` here already, this would allow us to get rid of `unwrap`s

ordian commented on ethcore/engines/validator-set/src/contract.rs:
```diff
@@ -48,31 +49,64 @@ pub struct ValidatorContract {
-			},
-			None => Err("No full client!".into()),
+	fn transact(&self, data: Bytes, gas_price: Option<U256>) -> Result<(), String> {
+		let client = self.client.read().as_ref().and_then(Weak::upgrade).ok_or("No client!")?;
```
do we need to upgrade the client ref here? it seems that we do that in `do_report_malicious` and `do_report_benign` already, so we could just pass it here, right?

afck commented:
I addressed the comments, rebased, and fixed the bug @varasev found: The nonce was not incremented between the randomness and the POSDAO calls after the refactoring (moving the transaction generation into separate methods).
Let's run [posdao-test-setup](https://github.com/poanetwork/posdao-test-setup/) as well before merging, to make sure.

Happy to discuss this PR in a call (see Discord)!

varasev commented:
> Let's run `posdao-test-setup` as well before merging, to make sure.

Yes, I will retest that

varasev commented:
> The nonce was not incremented between the randomness and the POSDAO calls after the refactoring (moving the transaction generation into separate methods). Let's run `posdao-test-setup` as well before merging, to make sure.

That works 👌

afck commented:
Works for me as well. :+1:

afck commented:
We made the changes we discussed yesterday. Please take a look at the two new commits.

varasev commented on ethcore/engines/validator-set/src/contract.rs:
```diff
@@ -165,13 +165,17 @@ impl ValidatorSet for ValidatorContract {
 		if let Err(s) = self.do_report_malicious(address, block, proof) {
 			warn!(target: "engine", "Validator {} could not be reported ({}) on block {}", address, s, block);
+		} else {
+			info!(target: "engine", "Reporting malicious validator {} on block {}", address, block);
```
Seems we already have the warning message inside the `do_report_malicious` function.

varasev commented on ethcore/engines/validator-set/src/contract.rs:
```diff
@@ -165,13 +165,17 @@ impl ValidatorSet for ValidatorContract {
 		if let Err(s) = self.do_report_benign(address, block) {
 			warn!(target: "engine", "Validator {} could not be reported ({}) on block {}", address, s, block);
+		} else {
+			info!(target: "engine", "Reporting validator {} on block {} (benign fault)", address, block);
```
Seems we already have the warning message inside the `do_report_benign` function.

afck commented:
You're right, @varasev, we completely missed that yesterday! I reverted that part of the change.

varasev commented:
@dvdplm Will it be correct to contribute to Parity docs through https://github.com/paritytech/wiki repository after our changes are released? Or there are other places we should add the docs to?

dvdplm commented:
> Will it be correct to contribute to Parity docs through https://github.com/paritytech/wiki repository after our changes are released? Or there are other places we should add the docs to?

That is a great question. I don't think we have a comprehensive plan for the client docs once we migrate away from the Parity org, so until we do I'd say the parity wiki is the least worst option?

dvdplm commented on ethcore/res/validator_contract.sol:
```diff
@@ -1,62 +1,103 @@
 // Source for the test AuRa validator set contract.
```
```suggestion
// Source for the test AuRa validator set contract. DO NOT USE IN PRODUCTION.
```

Maybe?

dvdplm commented on ethcore/res/validator_contract.sol:
```diff
@@ -66,39 +107,13 @@ contract TestList {
-				return false; // already reported
-			}
+		if (isValidatorBanned[_maliciousValidator]) {
+			// We shouldn't report of the malicious validator
```
```suggestion
			// We shouldn't report the malicious validator
```

varasev commented:
> I'd say the parity wiki is the least worst option?

Ok, thanks
